File: android/src/main/java/com/github/mustafaozhan/ccc/android/ui/calculator/CalculatorViewModel.kt
/*
 * Copyright (c) 2020 Mustafa Ozhan. All rights reserved.
 */

package com.github.mustafaozhan.ccc.android.ui.calculator

import androidx.lifecycle.viewModelScope
import com.github.mustafaozhan.ccc.android.base.BaseViewModel
import com.github.mustafaozhan.ccc.client.ui.calculator.CalculatorUseCase

class CalculatorViewModel(
    val useCase: CalculatorUseCase
) : BaseViewModel() {
    override fun onStart() {
        useCase.scope = viewModelScope
    }

    override fun onDestroy() {
        useCase.onDestroy()
    }
}

File: ios/ios/UI/CalculatorViewModel.swift
//
//  CalculatorViewModel.swift
//  ios
//
//  Created by Mustafa Ozhan on 25/12/2020.
//  Copyright © 2020 orgName. All rights reserved.
//

import Foundation
import Combine
import client

final class CalculatorViewModel: BaseViewModel {
    let useCase: CalculatorUseCase
    
    @Published var state: CalculatorState
    var effect = PassthroughSubject<CalculatorEffect, Never>()
    
    
    init(useCase: CalculatorUseCase) {
        self.useCase = useCase
        self.state = self.useCase.state
    }
    
    func startObserving() {
//        self.useCase.onChange(self.useCase.state as! Kotlinx_coroutines_coreFlow, provideNewState: { newState in
//            self.state = newState as! CalculatorState
//        })
        
        self.useCase.onChange(self.useCase.effect, provideNewState: { newState in
            self.effect.send(newState as! CalculatorEffect)
        })
    }

    func stopObserving() {
        self.useCase.onDestroy()
    }
}
